pkg/kubelet/kubelet_pods.go

```diff
@@ -129,7 +129,16 @@ func (kl *Kubelet) makeBlockVolumes(pod *v1.Pod, container *v1.Container, podVol
 }
 
 // makeMounts determines the mount points for the given container.
-func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, hostDomain, podIP string, podVolumes kubecontainer.VolumeMap, mounter mountutil.Interface) ([]kubecontainer.Mount, func(), error) {
+func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, hostDomain, podIP string, podVolumes kubecontainer.VolumeMap, mounter mountutil.Interface, expandEnvs []kubecontainer.EnvVar) ([]kubecontainer.Mount, func(), error) {
+
+	// Expand subpath variables
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeSubpath) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.VolumeSubpathExpansion) {
+		mounts := kubecontainer.ExpandContainerVolumeMounts(container, expandEnvs)
```

Comment:
Can this go even further down, like [here](https://github.com/kubernetes/kubernetes/pull/49388/files#diff-10055ae93a8699af13ceba0482fc43c3R177)?

Then you also don't need to generate a new array and can modify the field directly.

Reply:
This was the point I made about it being too late in the process at that section. It has to come before the actual making pf the mounts because the environment variables require expansion and then the mounts get created. If you put it in the spot where you suggested you can see in the kubelet logs that the container now has all the correct expansions but the mount still has $(POD_NAME) in the mount point.

Reply:
Yes. Strictly speaking this PR is about the container spec and not specifically about sub paths. So it has to do the expansion before it can perform the mounts. This then will evaluate any downstream api calls (same as container command and args) then construct the container spec, then effectively launch the container objects.

Reply:
My understanding is `makeMounts` constructs `[]kubecontainer.Mount` and after this, the container setup should not use `v1.Container.VolumeMounts` anymore.  Can you describe the error you are seeing if you don't update the `v1.Container.VolueMounts`?

Reply:
```

Reply:
So, if I modify the code slightly and put in the before and after picture, you will see that the container is not modified in the Expand. This was deliberately in keeping with the same principles applied to the container command and args expansion. I basically followed the same pattern

```
 mounts := kubecontainer.ExpandContainerVolumeMounts(container, expandEnvs)
        glog.V(4).Infof("Container Subpath Mount Points %+v\n", container.VolumeMounts)
        container.VolumeMounts = mounts
        glog.V(4).Infof("Expanded Subpath Mount Points %+v\n", mounts)
```
```
I0501 09:46:41.139032   25553 kubelet_pods.go:138] Container Subpath Mount Points [{Name:workdir1 ReadOnly:false MountPath:/logscontainer SubPath:$(POD_NAME) MountPropagation:<nil>} {Name:workdir2 ReadOnly:false MountPath:/testcontainer SubPath: MountPropagation:<nil>} {Name:default-token-hs6ch ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}]
```
```
I0501 09:46:41.139046   25553 kubelet_pods.go:140] Expanded Subpath Mount Points [{Name:workdir1 ReadOnly:false MountPath:/logscontainer SubPath:env-image MountPropagation:<nil>} {Name:workdir2 ReadOnly:false MountPath:/testcontainer SubPath: MountPropagation:<nil>} {Name:default-token-hs6ch ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}]
```